### PR TITLE
Emit events on remote db

### LINKED ISSUE #1414: Emit events on Remote DB
Currently remote db doesnt emit any events that db.EventListener expects.

https://github.com/NethermindEth/juno/blob/82e738efe30b7a748704330a24bd4811d59cb242/db/remote/db.go#L53-L55

### PR DESCRIPTION
**Pull request by RuneRogue**
So first I have added an listener field in DB struct then added code to emit event when a new transaction is created , view operation , update operation and when db is closed . I have also updated the WithListener function .

Fixed #1414

### CODE REVIEW

**Comment by PriyankMkwna** [comment]
I have made some changes that you have suggested.

File: db/remote/db.go [hunk 1]
```diff
@@ -13,11 +14,11 @@ import (
 var _ db.DB = (*DB)(nil)
 
 type DB struct {
-	ctx context.Context
-
+	ctx        context.Context
 	grpcClient *grpc.ClientConn
 	kvClient   gen.KVClient
 	log        utils.SimpleLogger
+	listener   db.EventListener
```
Comment (kirugan) [inline db/remote/db.go:21]: there are some improvements that we can do:
1. set `listener` to `SelectiveListener` in `New` function by default
2. remove nil checks for `d.listener`

File: db/remote/db.go [hunk 2]
```diff
@@ -48,10 +55,19 @@ func (d *DB) View(fn func(txn db.Transaction) error) error {
 }
 
 func (d *DB) Update(fn func(txn db.Transaction) error) error {
-	return db.Update(d, fn)
+	start := time.Now()
+
+	err := db.Update(d, fn)
+
+	if d.listener != nil {
+		d.listener.OnCommit(time.Since(start))
```
Comment (kirugan) [inline db/remote/db.go:63]: after removing `d.listener` check consider using `defer func() { d.listener.OnCommit(...)} ` as in the rest of codebase (but only here)

**Review by kirugan — CHANGES_REQUESTED** [review]
get rid of nil checks + add defer in 1 place

**Comment by PriyankMkwna** [comment]
Updated dg.go as you have suggested.

**Review by kirugan — APPROVED** [review]

**Review by AnkushinDaniil — APPROVED** [review]

**Review by rianhughes — APPROVED** [review]

**Comment by kirugan** [comment]
@RuneRogue it's almost done: there is one minor change - see linter issues:
```
db/remote/db.go:31: File is not `gci`-ed with --skip-generated -s standard -s default (gci)
		OnIOCb:     func(write bool, duration time.Duration) {}, 
db/remote/db.go:51: File is not `gci`-ed with --skip-generated -s standard -s default (gci)
	
make: *** [Makefile:89: lint] Error 1
```
you can verify it locally with `make lint`

**Comment by kirugan** [comment]
@RuneRogue thanks for contribution. Please check OnlyDust

**Pull request merged**